src/merkle/simple_smt/mod.rs

```diff
@@ -186,27 +194,21 @@ impl SimpleSmt {
     // STATE MUTATORS
     // --------------------------------------------------------------------------------------------
 
-    /// Replaces the leaf located at the specified key, and recomputes hashes by walking up the
-    /// tree.
+    /// Updates value of the leaf at the specified index returning the old leaf value.
+    ///
+    /// This also recomputes all hashes between the leaf and the root, updating the root itself.
     ///
     /// # Errors
-    /// Returns an error if the specified key is not a valid leaf index for this tree.
-    pub fn update_leaf(&mut self, key: u64, value: Word) -> Result<(), MerkleError> {
-        let index = NodeIndex::new(self.depth(), key)?;
-        if !self.check_leaf_node_exists(key) {
-            return Err(MerkleError::NodeNotInSet(index.value()));
-        }
-        self.insert_leaf(key, value)?;
-
-        Ok(())
-    }
+    /// Returns an error if the index is greater than the maximum tree capacity, that is 2^{depth}.
+    pub fn update_leaf(&mut self, index: u64, value: Word) -> Result<Word, MerkleError> {
```

Comment:
Something to consider: we often expect the new root when we mutate a tree. This would avoid having to do something like this

```rust
tree.update_leaf(index, value)?;
let root = tree.root();
```

I don't see a clear case when the user will need to fetch the previous value - perhaps only to check if it was empty? Given this would be an exception, maybe that case would require an additional call

```rust
let previous = tree.get_leaf(index)?;
let root = tree.update_leaf(index, value)?;
```

Reply:
Yeah - either approach could be valid. We decided to go with returning the previous to keep it consistent with `mtree_get` instruction.